extension/socks5proxy/bin/windowsbypass.cpp

```diff
@@ -0,0 +1,351 @@
+                              MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+                              (LPWSTR)&buffer, 0, nullptr);
+  QString result = QString::fromWCharArray(buffer, size);
+  LocalFree(buffer);
+  return result;
+}
+
+quint64 WindowsBypass::getVpnLuid() const {
+  // Get the LUID of the wireguard interface, if it's up.
+  NET_LUID luid;
+  GUID vpnInterfaceGuid = WIREGUARD_NT_GUID;
+  if (ConvertInterfaceGuidToLuid(&vpnInterfaceGuid, &luid) != NO_ERROR) {
+    return 0;
+  } else {
+    return luid.Value;
+  }
+}
+
+// Refresh our understanding of the current interfaces, and identify the VPN
+// tunnel, if running.
+void WindowsBypass::refreshIfMetrics() {
+  // Fetch the interface table.
+  MIB_IPINTERFACE_TABLE* table;
+  DWORD result = GetIpInterfaceTable(AF_UNSPEC, &table);
+  if (result != NO_ERROR) {
+    qWarning() << "GetIpInterfaceTable() failed:" << win32strerror(result);
+    return;
+  }
+  auto guard = qScopeGuard([table]() { FreeMibTable(table); });
+
+  const quint64 vpnInterfaceLuid = getVpnLuid();
+  for (ULONG i = 0; i < table->NumEntries; i++) {
+    const MIB_IPINTERFACE_ROW* row = &table->Table[i];
+    if (row->InterfaceLuid.Value == vpnInterfaceLuid) {
+      continue;
+    }
+
+    if (!m_interfaceData.contains(row->InterfaceLuid.Value)) {
+      // Ignore interfaces with no source addresses.
+      continue;
+    }
+    m_interfaceData[row->InterfaceLuid.Value].metric = row->Metric;
+  }
+}
+
+void WindowsBypass::refreshAddresses() {
+  // Get the unicast address table.
+  MIB_UNICASTIPADDRESS_TABLE* table;
+  DWORD result = GetUnicastIpAddressTable(AF_UNSPEC, &table);
+  if (result != NO_ERROR) {
+    qWarning() << "GetUnicastIpAddressTable() failed:" << win32strerror(result);
+    return;
+  }
+  auto guard = qScopeGuard([table]() { FreeMibTable(table); });
+
+  // Clear the inteface data while we rebuild it.
+  m_interfaceData.clear();
+
+  // Populate entries.
+  const quint64 vpnInterfaceLuid = getVpnLuid();
+  for (ULONG i = 0; i < table->NumEntries; i++) {
+    const MIB_UNICASTIPADDRESS_ROW* row = &table->Table[i];
+    if (row->SkipAsSource) {
+      continue;
+    }
+    if (row->InterfaceLuid.Value == vpnInterfaceLuid) {
+      continue;
+    }
+    if ((row->DadState == IpDadStateInvalid) ||
+        (row->DadState == IpDadStateTentative) ||
+        (row->DadState == IpDadStateDuplicate)) {
+      continue;
+    }
+
+    QHostAddress addr;
+    if (row->Address.si_family == AF_INET) {
+      quint32 rowAddr = row->Address.Ipv4.sin_addr.s_addr;
+      addr.setAddress(qFromBigEndian<quint32>(rowAddr));
+    } else if (row->Address.si_family == AF_INET6) {
+      addr.setAddress(row->Address.Ipv6.sin6_addr.s6_addr);
+    } else {
+      continue;
+    }
+
+    // Ignore unrouteable addresses.
+    if (addr.isLinkLocal() || addr.isMulticast() || addr.isLoopback()) {
+      continue;
+    }
+
+    // Store the address.
+    // TODO: Compare amongst multiple addresses on an interface.
```

Comment:
This TODO actually leads to breakage of IPv6 on my device because it winds up selecting a non-routeable IPv6 source address. So, this isn't really fit for purpose until it gets addressed.

Reply:
I added a bit more filtering around address selection to prevent using local addresses and it seems to have gotten IPv6 working again. I'm going to leave this TODO in here for a bit longer, because I think we could still do better in terms of source address selection, but it seems to be perfectly workable as-is now.

Reply:
As a note, there are two dimensions to consider here for doing proper IPv6 source address selection.

First, we could join a network which has multiple global prefixes and to do selection correctly, we are supposed to choose the source address which best matches the destination address. This would require storing multiple addresses and doing that selection later at socket creation.

Second, Windows will typically create multiple IPv6 addresses for each prefix, but with multiple address suffixes. We only need to store one of these, and we probably want the address with a randomized suffix, as this helps preserve the users privacy by preventing tracking of the user by their MAC address (which can be leaked in the suffix).